cmd/k8srules/controller.go

```diff
@@ -0,0 +1,273 @@
+	}
+
+	config.GroupVersion = &groupversion
+	config.APIPath = "/apis"
+	config.ContentType = runtime.ContentTypeJSON
+	config.NegotiatedSerializer = serializer.DirectCodecFactory{CodecFactory: api.Codecs}
+
+	schemeBuilder := runtime.NewSchemeBuilder(
+		func(scheme *runtime.Scheme) error {
+			scheme.AddKnownTypes(
+				groupversion,
+				&kuberules.Rule{},
+				&kuberules.RuleList{},
+				&api.ListOptions{},
+				&api.DeleteOptions{},
+			)
+			return nil
+		})
+	schemeBuilder.AddToScheme(api.Scheme)
+
+	client, err := rest.RESTClientFor(config)
+	if err != nil {
+		log.Printf("Failed to create the in-cluster client. %s\n", err)
+		return nil, err
+	}
+
+	return client, nil
+}
+
+func (c *controller) validateRule(rule *kuberules.Rule) error {
+	var updRule kuberules.Rule
+	var result kuberules.Rule
+
+	log.Printf("Rule %s is being validated\n", rule.Metadata.Name)
+
+	if rule.Spec.ID == "" {
+		updRule.Spec.ID = rule.Metadata.Name
+	}
+
+	err := c.validator.Validate(rule.Spec)
+	if err != nil {
+		log.Printf("Rule %#v is not valid. %s\n", rule, err)
+		updRule.Status.State = kuberules.RuleStateInvalid
+		updRule.Status.Message = err.Error()
+	} else {
+		log.Printf("Rule %s is valid\n", rule.Metadata.Name)
+		updRule.Status.State = kuberules.RuleStateValid
+	}
+
+	data, err := json.Marshal(&updRule)
+	if err != nil {
+		log.Printf("Failed to encode status for rule %s. %s\n", rule.Metadata.Name, err)
+		return err
+	}
+
+	err = c.client.Patch(api.MergePatchType).
+		Resource(fmt.Sprintf("%ss", kuberules.ResourceName)).
+		Namespace(rule.Metadata.Namespace).
+		Name(rule.Metadata.Name).
+		Body(data).
+		Do().
+		Into(&result)
+
+	if err != nil {
+		log.Printf("Failed to update status for rule %s. %s\n", rule.Metadata.Name, err)
+	}
+
+	return err
+}
+
+// addRule is the callback invoked by the Kubernetes cache when a rule API resource is added.
+func (c *controller) addRule(obj interface{}) {
+	rule, ok := obj.(*kuberules.Rule)
+	if !ok {
+		log.Printf("Invalid rule added: object is of type %T\n", obj)
+		return
+	}
+
+	log.Printf("Rule object added: %s (state: %s, version: %s)\n",
+		rule.Metadata.Name,
+		rule.Status.State,
+		rule.Metadata.ResourceVersion)
+
+	if rule.Status.State == "" {
+		c.validateRule(rule)
+	}
+}
+
+// updateRule is the callback invoked by the Kubernetes cache when a rule API resource is updated.
+func (c *controller) updateRule(oldObj, newObj interface{}) {
+	oldRule, ok := oldObj.(*kuberules.Rule)
+	if !ok {
+		log.Printf("Invalid rule update: old object is of type %T\n", oldObj)
+		return
+	}
+	newRule, ok := newObj.(*kuberules.Rule)
+	if !ok {
+		log.Printf("Invalid rule update: new object is of type %T\n", newObj)
+		return
+	}
+
+	log.Printf("Rule object updated: %s (state: %s, version: %s -> %s)\n",
+		newRule.Metadata.Name,
+		newRule.Status.State,
+		oldRule.Metadata.ResourceVersion,
+		newRule.Metadata.ResourceVersion)
+
+	if newRule.Status.State == "" {
+		c.validateRule(newRule)
+	}
+}
```

Comment:
@rshriram Only the rule instance that was deleted. Not the entire array of rules.